docs/proposals/21-ipv6-singlestack-local.md

```diff
@@ -0,0 +1,287 @@
+
+## Proposal
+
+A new feature gate `IPv6SingleStack` is added to gardener-apiserver.
+The IPv6-related fields and values can only be used if the feature gate is enabled.
+
+The feature gate serves the purpose of disabling the feature in productive Gardener installations and prevents users from configuring IPv6 networking for their shoot clusters, while it is still under development and not supported on cloud infrastructure.
+As part of this enhancement, the feature gate is supposed to be enabled only in the local environment.
+Later on, the feature gate may also be used for safe-guarding maturing and enablement of a IPv6 single-stack implementation on cloud infrastructure.
+The feature gate is not supposed to be toggled back and forth.
+
+### `Shoot` API
+
+The `Shoot.spec.networking` section is extended as follows:
+
+```yaml
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+spec:
+  networking:
+    pods: 2001:db8:1::/48
+    nodes: 2001:db8:2::/48
+    services: 2001:db8:3::/48
+    # ...
+    ipFamilies:
+    - IPv6
+```
+
+`ipFamilies` is the central setting for specifying the IP families used for shoot networking.
+This field is inspired by the `Service.spec.ipFamilies` field in Kubernetes ([doc](https://kubernetes.io/docs/concepts/services-networking/dual-stack/#services)).
+The default value is `["IPv4"]` (IPv4 single-stack).
+If the `IPv6SingleStack` feature gate is enabled, `["IPv6"]` can be specified to switch to IPv6 single-stack.
+
+Later on, `["IPv4","IPv6"]` or `["IPv6","IPv4"]` can be supported for dual-stack networking.
+In dual-stack networking, the ordering of `ipFamilies` is needed for correctly configuring Kubernetes components, e.g., the API server's `--service-cluster-ip-range` flag controls which IP family is used to allocate the primary `clusterIP` of services.
+
+Instead of explicitly configuring the `ipFamilies` field, the used families could be determined implicitly from the corresponding CIDR fields.
+However, when using IPv6 single-stack networking on cloud infrastructure, users might want to use IPv6 prefixes assigned to them by the provider.
+Typically, these are not known by the user upfront and need to be allocated during cluster creation.
+In this case, users might not supply any of the CIDRs in their shoot specification.
+Hence, we need a central field that allows configuring the used IP families explicitly instead of implicitly.
+That's exactly the purpose of the `ipFamilies` field.
+
+If IPv6 single-stack networking is configured via the `ipFamilies` field, the existing `Shoot.spec.networking.{pods,services,nodes}` fields are used to specify the IPv6 CIDRs instead of the IPv4 CIDRs.
+
+All gardener components and extensions need to respect the `ipFamilies` field and handle it correctly, e.g., in API validation, defaulting, and configuring Shoot components.
```

Comment:
`Service.spec.ipFamilies` is already a pretty bad example, as it interacts with many other settings in a very complex way.
Why not auto-detect the setting via the ranges and add nemonics for "cloud provider assigned IPv4/IPv6" [like we did in our first  GEP attempt](https://github.com/AndreBorrmann/gep_wip/blob/main/18-dualstack_ipv4_ipv6_seed_shoot.md#configuration)?

Reply:
> It could be added as a new default value or the empty value could also mean `IPv4`. Which approach do you want to take? Please be explicit in this regard.

Generally speaking, new API fields should be defaulted by the API server to their implicit value – in this case, `["IPv4"]`. 
Empty values are less explicit and require client-side knowledge of the API semantics, which is an anti-pattern because it is bad for versioning and compatibility.
We made the respective description more explicit in this regard.